File: 05.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script

{-# OPTIONS_GHC -Wall #-}

import Data.Char (toUpper)

reactive :: Char -> Char -> Bool
reactive a b = (a /= b) && (toUpper a == toUpper b)

react :: [Char] -> [Char]
react = go [] where
    go rwd [] = rwd
    go [] (f : fwd) = go [f] fwd
    go (r : rwd) (f : fwd)
        | reactive r f = go rwd fwd
        | otherwise = go (f : r : rwd) fwd

remove :: Char -> [Char] -> [Char]
remove x = filter (\c -> toUpper c /= toUpper x)

main :: IO ()
main = do
    input <- readFile "05.input"
    let polymer = head $ lines input
    -- part 1
    let reacted = react polymer
    print $ length reacted
    -- part 2
    print $ minimum [length $ react $ remove x polymer | x <- ['a'..'z']]

File: 05.py
#!/usr/bin/env python3

import string


def reactive(a, b):
    return a != b and a.upper() == b.upper()


def react(chars):
    ret = []
    for c in chars:
        if ret and reactive(ret[-1], c):
            ret.pop()
        else:
            ret.append(c)
    return ret


def remove(x, chars):
    return filter(lambda c: c not in {x.lower(), x.upper()}, chars)


if __name__ == '__main__':
    with open('05.input') as f:
        polymer = list(f.read().strip())

    # part 1
    print(len(react(polymer)))

    # part 2
    print(min(len(react(remove(x, polymer))) for x in string.ascii_lowercase))
